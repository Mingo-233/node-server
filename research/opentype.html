<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
    <style>
        body {
            /* margin: 0; */
            border: 1px solid black;
        }

        #preview {
            /* position: absolute;
        top: 100px;
        right: 100px; */
            /* margin: 30px 100px; */
            margin-top: 300px;
            border: 1px solid red;
        }

        .d-hotline.h-btn {
            border-radius: 50% !important;
            border: 1px solid #fff !important;
            box-shadow: none !important;
            width: 48px !important;
            height: 48px !important;
        }

        .d-hotline.h-btn #chat-icon {
            top: 30% !important;
        }
    </style>
</head>

<body>
    <div id="preview">

    </div>
    <!-- <script src="https://opentype.js.org/dist/opentype.js"></script> -->
    <script src="./opentype.js"></script>
    <script src="https://unpkg.com/wawoff2@2.0.1/build/decompress_binding.js"></script>

    <script>
        function load() {
            // const buffer = document.getElementById('myfile').files[0].arrayBuffer();

            // const buffer = fetch("https://cdn.pacdora.com/font_v2/DM_Sans/DMSans-Bold.woff2").then((res) =>
            // const buffer = fetch("https://cdn.pacdora.com/font_v2/Anton/Anton-Regular.woff2").then((res) =>
            // const buffer = fetch("https://yun.baoxiaohe.com/font/05345406-5b37-4be2-a689-0360185bc5ac.woff2").then((res) =>
            const buffer = fetch("https://oss.pacdora.cn/frontend/webFonts/HarmonyOS_Sans_SC_Regular.woff2").then((res) =>

                // const buffer = fetch("https://cdn.pacdora.com/font_v2/Pacifico/Pacifico-Regular.woff2").then((res) =>

                res.arrayBuffer()
            );
            buffer.then((data) => {
                const decompressed = Module.decompress(data);
                const app = opentype.parse(decompressed);
                console.log('app', app);
                // let glyphs = Object.keys(app.glyphs.glyphs)

                // let fontGlyphsName = []
                // for (const element of glyphs) {
                //     let name = app.glyphs.glyphs[element].name
                //     if (name.includes('quote')) {
                //         console.log('123123');

                //         console.log(element);

                //     }
                //     fontGlyphsName.push(name)
                // }
                // console.log('fontGlyphsName', fontGlyphsName);
                // let str = "abcdef"
                let str = "abcdefabcdef"
                console.log('unicode', str.codePointAt())
                // const result = app.charToGlyph(str)
                // console.log(result);

                const path = app.getPath(str, 0, 0, 4, {
                    hinting: false
                });

                const svgPath = path.toSVG(3)

                const pathBoundingBox = path.getBoundingBox()
                console.log(pathBoundingBox);
                const viewBox = `${pathBoundingBox.x1} ${pathBoundingBox.y1} 20 20`
                console.log(svgPath);
                const svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg')
                svg.setAttribute('viewBox', viewBox)
                svg.setAttribute('width', '20mm')
                svg.setAttribute('height', '20mm')
                svg.setAttribute('xmlns', 'http://www.w3.org/2000/svg')
                svg.setAttribute('stroke', 'none')
                svg.innerHTML = svgPath
                document.getElementById('preview').appendChild(svg)
                // const font = opentype.parse(data);


            });
        }
        window.load = load;

        function downloadSvg() {
            const svg = document.getElementById('preview').innerHTML
            const blob = new Blob([svg], { type: 'image/svg+xml' });
            const url = URL.createObjectURL(blob);
            const a = document.createElement('a');
            a.href = url;
            a.download = 'font.svg';
            a.click();
        }
        window.downloadSvg = downloadSvg;
        setTimeout(() => {
            load();
        }, 1000);
    </script>
</body>

</html>