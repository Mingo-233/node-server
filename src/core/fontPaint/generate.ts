// @ts-nocheck
import { createElement } from "@/nodes/index";

export function genSvgCode(pathWhole, config) {
  const {
    position,
    isVertical,
    pathPartsTransform,
    pathPartsAlignTransform,
    domBoxSize,
    hasCnChar,
    adobeAiTransform = "",
    DPI = 1,
  } = config;
  let svgPathString = "";
  //   判断是个对象
  if (Object.prototype.toString.call(pathWhole) === "[object Object]") {
    // console.log(
    //   "paths",
    //   pathWhole,
    //   pathPartsTransform,
    //   pathPartsAlignTransform
    // );
    const lines = Object.keys(pathWhole);

    for (let i = 0; i < lines.length; i++) {
      let paths = pathWhole[lines[i]];
      const currentLineTransform = pathPartsTransform[i];
      const currentAlignTransform = pathPartsAlignTransform[i];
      paths.forEach((path, index) => {
        if (!path) return;
        let pathString = path.toSVG(6);
        const pathTransform = currentLineTransform[index];
        if (pathTransform) {
          pathString = `${pathString.slice(
            0,
            5
          )} transform="${pathTransform}" ${pathString.slice(5)}`;
        }
        const alignTransform = currentAlignTransform[index];
        const template = `
                  <g transform="${alignTransform ? alignTransform : ""}">
                      ${pathString}
                  </g>
                  `;
        svgPathString += template;
      });
    }
  }

  // const svgDom = document.createElementNS("http://www.w3.org/2000/svg", "svg");
  let svgDom = "";
  let G_Template = "";
  //   中英文都存在垂直情况
  if (isVertical && hasCnChar) {
    G_Template = `
      <g >
      ${svgPathString}
      </g>
      `;
    svgDom = createElement(
      "svg",
      {
        xmlns: "http://www.w3.org/2000/svg",
        id: "preview",
        width: domBoxSize.width + "mm",
        height: domBoxSize.height + "mm",
        viewBox: `${position.x1} ${position.y1} ${
          position.x2 + domBoxSize.width
        } ${position.y2 + domBoxSize.height}`,
        transform: `${adobeAiTransform}`,
        fill: "red",
      },
      G_Template
    );
  } else if (isVertical) {
    let originWidth = domBoxSize.width;
    let originHeight = domBoxSize.height;
    // 在Ai中
    const _transform = `${adobeAiTransform}  rotate(90)
translate(0,-${domBoxSize.width * DPI})
      `;
    const G_Transform = ``;
    G_Template = `
      <g ${G_Transform}>
      ${svgPathString}
      </g>
      `;
    svgDom = createElement(
      "svg",
      {
        xmlns: "http://www.w3.org/2000/svg",
        id: "preview",
        width: domBoxSize.height + "mm",
        height: domBoxSize.width + "mm",
        viewBox: `${position.x1} ${position.y1} ${originHeight} ${originWidth}`,
        transform: `${_transform}`,
        fill: "red",
      },
      G_Template
    );
  } else {
    const G_Transform = ``;
    G_Template = `
      <g ${G_Transform}>
      ${svgPathString}
      </g>
      `;
    svgDom = createElement(
      "svg",
      {
        xmlns: "http://www.w3.org/2000/svg",
        id: "preview",
        width: domBoxSize.width + "mm",
        height: domBoxSize.height + "mm",
        viewBox: `${position.x1} ${position.y1} ${
          position.x2 + domBoxSize.width
        } ${position.y2 + domBoxSize.height}`,
        transform: `${adobeAiTransform}`,
        fill: "red",
      },
      G_Template
    );
  }

  return svgDom;
}
