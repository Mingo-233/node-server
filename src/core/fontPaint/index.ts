export { parseText } from "./parse";
export { transformText } from "./transform";
export { genTextSvg } from "./generate";
export { matchSymbol, isCharSupported, getDefaultFontApp } from "./utils";
