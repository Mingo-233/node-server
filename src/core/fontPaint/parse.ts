// @ts-nocheck
const pathPartType = {
  zh: 1,
  en: 2,
};

function createWordPathContext() {
  const context = {
    word: "",
    line: 0, //层数
    pathParts: {
      0: [],
    },
    pathPartsTransform: {
      0: [],
    },
    pathPartsAlignTransform: {
      0: [],
    },
    resetWord() {
      context.word = "";
    },
    addWord(text) {
      context.word += text;
    },
    backWord() {
      context.word = context.word.slice(0, -1);
    },
    addPath(path) {
      context.pathParts[context.line].push(path);
    },
    nextLine() {
      context.line = context.line + 1;
      context.pathParts[context.line] = [];
      context.pathPartsTransform[context.line] = [];
      context.pathPartsAlignTransform[context.line] = [];
    },
    addTransform(transform) {
      context.pathPartsTransform[context.line].push(transform);
    },
    addAlignTransform(transform) {
      context.pathPartsAlignTransform[context.line].push(transform);
    },
  };
  return context;
}
function getPath(fontApp, text, fontSize) {
  return fontApp.getPath(text, 0, 0, fontSize);
}
export function getTextPaths(fontApp, config) {
  if (config.vertical && hasCnWords(config.text)) {
    return getVerticalTextPaths(fontApp, config);
  }
  const textArr = config.text.split("");
  console.log("字符拆分 en", textArr);
  const context = createWordPathContext();
  const position = {
    x1: 0,
    y1: 0,
    x2: 0,
    y2: 0,
  };
  const svgSize = {
    width: 0,
    height: 0,
  };
  const lineHeightRatio = config.textLineHeight / config.fontSize;
  let lineHeightTop = 0;
  let lineHeight = 0;
  const isVertical = !!config.vertical;
  //   编辑器中的选框大小
  const MAX_WIDTH = isVertical ? config.MaxHeight : config.MaxWidth;
  const MAX_HIGHT = isVertical ? config.MaxWidth : config.MaxHeight;
  for (let i = 0; i < textArr.length; i++) {
    const text = textArr[i];
    if (text === "\n") {
      settlePath(context, false);
      context.nextLine();
      // 跳过 换行字符的处理
      // i++;
      continue;
    }
    context.addWord(text);
    const path = getPath(fontApp, context.word, config.fontSize);
    const pathBoundingBox = path.getBoundingBox();
    if (i === 0) {
      position.x1 = pathBoundingBox.x1;
      position.y1 = pathBoundingBox.y1;
    }
    const currentWidth = pathBoundingBox.x2 - pathBoundingBox.x1;
    if (currentWidth > svgSize.width) {
      svgSize.width = currentWidth;
    }
    const currentHeight = pathBoundingBox.y2 - pathBoundingBox.y1;
    if (currentHeight > svgSize.height) {
      svgSize.height = currentHeight;
      lineHeight = currentHeight * lineHeightRatio;
      lineHeightTop = (lineHeight - currentHeight) / 2;
    }
    // 超出选框边界
    if (pathBoundingBox.x2 > MAX_WIDTH) {
      context.backWord();
      settlePath(context, false);
      context.addWord(text);
      context.nextLine();
    }
    //   end 最后一个字符
    if (i === textArr.length - 1) {
      settlePath(context, false);
    }
  }
  //   结算当前路径
  function settlePath(context, isBreak = true) {
    const path = getPath(fontApp, context.word, config.fontSize);
    const pathBoundingBox = path.getBoundingBox();
    const transform = computedAlignTransform(
      config.textAlign,
      MAX_WIDTH,
      MAX_HIGHT,
      pathBoundingBox
    );
    context.addPath(path);
    context.addAlignTransform(transform);
    const translateY = lineHeight * context.line + lineHeightTop;
    const pathTransform = `translate(0,${translateY})`;
    context.addTransform(pathTransform);

    context.resetWord();
    if (isBreak) {
      let currentHeight = position.y2 - position.y1;
      svgSize.height = svgSize.height + currentHeight;
    }
  }
  function computedAlignTransform(textAlign, width, height, pathBoundingBox) {
    const pathWidth = pathBoundingBox.x2 - pathBoundingBox.x1;
    let translateX;
    switch (textAlign) {
      case "center":
        translateX = (width - pathWidth) / 2;
        return `translate(${translateX},0)`;
      case "right":
        translateX = width - pathWidth;
        return `translate(${translateX},0)`;

      default:
        return "";
    }
  }
  return {
    pathParts: context.pathParts,
    pathPartsAlignTransform: context.pathPartsAlignTransform,
    pathPartsTransform: context.pathPartsTransform,
    position,
    svgSize,
    lineHeight,
    isVertical,
    hasCnChar: false,
    domBoxSize: {
      width: config.MaxWidth,
      height: config.MaxHeight,
    },
  };
}

function getVerticalTextPaths(fontApp, config) {
  const context = createWordPathContext();
  const position = {
    x1: 0,
    y1: 0,
    x2: 0,
    y2: 0,
  };
  const svgSize = {
    width: 0,
    height: 0,
  };
  let lineHeight = 26.46;
  const isVertical = !!config.vertical;
  //   编辑器中的选框大小
  const MAX_WIDTH = config.MaxWidth;
  const MAX_HIGHT = config.MaxHeight;

  const { textInfoArr, maxItemHeight, maxItemWidth, line } = mapText(
    fontApp,
    config
  );
  console.log("line", line);
  const { lineNum, breakLineIndex, maxContentHeightArr } = line;
  let currentLine = 1;
  let colHandleNum = 0;
  textInfoArr.forEach((textInfo, index) => {
    if (currentLine > lineNum) return;
    if (breakLineIndex.includes(index)) {
      // 当前命中换行索引
      context.nextLine();
      currentLine++;
      colHandleNum = 0;
    }
    if (!textInfo.path) return;
    context.addPath(textInfo.path);
    if (textInfo.type === pathPartType.zh) {
      const translateX = MAX_WIDTH - maxItemWidth * currentLine;
      //当前这列上这个字是第几个,计算垂直方向偏移量
      const translateY = maxItemHeight * colHandleNum;
      const transform = `translate(${translateX},${translateY})`;
      context.addTransform(transform);
      const alignTransform = computedPathTransform(
        config.textAlign,
        maxContentHeightArr[currentLine - 1]
      );
      context.addAlignTransform(alignTransform);
      colHandleNum++;
    } else {
      console.log("英文", textInfo);
      const translateX = MAX_WIDTH - maxItemWidth * currentLine;
      //当前这列上这个字是第几个,计算垂直方向偏移量
      const translateY = maxItemHeight * colHandleNum;
      // 先平移到目标位置 然后绕svg左上角圆点旋转90度，然后再平移修复旋转移动的距离
      const transform = `translate(${translateX},${translateY}) rotate(90,${position.x1},${position.y1}) translate(0,-${maxItemWidth})`;
      context.addTransform(transform);
      const alignTransform = computedPathTransform(
        config.textAlign,
        maxContentHeightArr[currentLine - 1]
      );
      context.addAlignTransform(alignTransform);
      colHandleNum++;
    }
  });
  //   计算整体的偏移量
  function computedPathTransform(textAlign, maxContentHeigh) {
    switch (textAlign) {
      case "center":
        offsetY = (MAX_HIGHT - maxContentHeigh) / 2;
        return `translate(0,${offsetY})`;
      case "right":
        offsetY = MAX_HIGHT - maxContentHeigh;
        return `translate(0,${offsetY})`;

      default:
        return "";
    }
  }
  function mapText(fontApp, config) {
    const textInfoArr = [];
    const textArr = config.text.split("");
    console.log("字符拆分", textArr);

    // 每列都水平位移，要找出最大的宽度
    let maxItemWidth = 0;

    // 每个字都垂直位移，要找出最大的高度
    let maxItemHeight = 0;
    // 已经识别的索引
    let identifiedIndex = -9999;
    for (let i = 0; i < textArr.length; i++) {
      if (i <= identifiedIndex) {
        continue;
      }
      const textItem = textArr[i];
      if (textItem === "\n") {
        textInfoArr.push({
          path: null,
          pathBoundingBox: {},
          text: textItem,
          isBreak: true,
          type: pathPartType.zh,
        });
        continue;
      }
      if (isChinese(textItem)) {
        const path = getPath(fontApp, textItem, config.fontSize);
        const pathBoundingBox = path.getBoundingBox();
        const currentPathWidth = pathBoundingBox.x2 - pathBoundingBox.x1;
        if (currentPathWidth > maxItemWidth) {
          maxItemWidth = currentPathWidth;
        }
        const currentPathHeight = pathBoundingBox.y2 - pathBoundingBox.y1;
        if (currentPathHeight > maxItemHeight) {
          maxItemHeight = currentPathHeight;
        }

        textInfoArr.push({
          path,
          pathBoundingBox,
          text: textItem,
          isBreak: false,
          type: pathPartType.zh,
        });
        if (i === 0) {
          position.x1 = pathBoundingBox.x1;
          position.y1 = pathBoundingBox.y1;
        }
      } else {
        const result = identifyNext(textItem, i, textArr);
        identifiedIndex = result.lastIndex;
        // 因为英文要旋转过来，所以它的宽度就是高度
        const currentPathHeight =
          result.pathBoundingBox.x2 - result.pathBoundingBox.x1;

        textInfoArr.push({
          path: result.path,
          pathBoundingBox: result.pathBoundingBox,
          text: result.text,
          isBreak: false,
          type: pathPartType.en,
          height: currentPathHeight,
          chilePath: result.chilePath,
        });
        if (i === 0) {
          position.x1 = result.pathBoundingBox.x1;
          position.y1 = result.pathBoundingBox.y1;
        }
      }
    }
    const line = computedLine(textInfoArr, maxItemHeight);

    return { textInfoArr, maxItemWidth, maxItemHeight, line };
  }
  function computedLine(textInfoArr, height) {
    let lineNum = 1;
    // 换行标记索引
    let breakLineIndex = [];
    let maxContentHeightArr = [];
    let accumulatorPathHeight = 0;
    console.log("textInfoArr", textInfoArr);
    for (let index = 0; index < textInfoArr.length; index++) {
      const textInfo = textInfoArr[index];
      if (textInfo.isBreak) {
        doBreak(index);
        continue;
      }
      const _height = textInfo.height || height;
      accumulatorPathHeight = accumulatorPathHeight + _height;
      if (accumulatorPathHeight > MAX_HIGHT) {
        if (textInfo.chilePath) {
          let preAccumulatorPathHeight = accumulatorPathHeight - _height;
          let lastPath = textInfo.chilePath[textInfo.chilePath.length - 1];
          let lastPathHeight =
            lastPath.path.getBoundingBox().x2 -
            lastPath.path.getBoundingBox().x1;
          console.log("while");
          // 从children找到能刚好能排列下的子路径
          while (preAccumulatorPathHeight + lastPathHeight > MAX_HIGHT) {
            textInfo.chilePath.pop();
            lastPath = textInfo.chilePath[textInfo.chilePath.length - 1];
            lastPathHeight =
              lastPath.path.getBoundingBox().x2 -
              lastPath.path.getBoundingBox().x1;
          }
          console.log("textInfo.chilePath", textInfo.chilePath);
          //   原本的path 拆分成2个
          const newFirstTextInfo = {
            ...textInfo.chilePath[textInfo.chilePath.length - 1],
            type: pathPartType.en,
          };
          newFirstTextInfo.pathBoundingBox =
            newFirstTextInfo.path.getBoundingBox();
          const newSecondText = textInfo.text.slice(
            newFirstTextInfo.text.length
          );
          const newSecondTextInfoPath = getPath(
            fontApp,
            newSecondText,
            config.fontSize
          );
          const newSecondTextInfoBoundingBox =
            newSecondTextInfoPath.getBoundingBox();
          const newSecondTextInfo = {
            text: newSecondText,
            path: newSecondTextInfoPath,
            type: pathPartType.en,
            pathBoundingBox: newSecondTextInfoBoundingBox,
            height:
              newSecondTextInfoBoundingBox.x2 - newSecondTextInfoBoundingBox.x1,
          };

          const newPaths = [newFirstTextInfo, newSecondTextInfo];
          textInfoArr.splice(index, 1, ...newPaths);
          console.warn("preAccumulatorPathHeight", preAccumulatorPathHeight);
          accumulatorPathHeight = preAccumulatorPathHeight + lastPathHeight;
          doBreak(index + 1);

          continue;
        } else {
          accumulatorPathHeight = accumulatorPathHeight - _height;
          doBreak(index);
          //   最后一个字符，高度即它本身
          if (index === textInfoArr.length - 1) {
            maxContentHeightArr.push(_height);
          }
          continue;
        }
      }
    }
    if (maxContentHeightArr.length !== lineNum) {
      maxContentHeightArr.push(accumulatorPathHeight);
    }
    function doBreak(index) {
      lineNum++;
      maxContentHeightArr.push(accumulatorPathHeight);
      console.warn("maxContentHeightArr", maxContentHeightArr);
      accumulatorPathHeight = 0;
      breakLineIndex.push(index);
    }
    return { lineNum, breakLineIndex, maxContentHeightArr };
  }
  //   结算当前路径
  function settlePath() {}

  function identifyNext(char, index, textInfoArr) {
    let nextText = textInfoArr[index + 1];
    let path = null;
    const chilePath = [];
    while (!isEnd(nextText)) {
      index++;
      char = `${char}${nextText}`;
      nextText = textInfoArr[index + 1];
      path = getPath(fontApp, char, config.fontSize);
      chilePath.push({
        text: char,
        path: path,
      });
    }
    if (isEnglish(char)) {
      path = getPath(fontApp, char, config.fontSize);
    }
    const pathBoundingBox = path.getBoundingBox();
    return {
      lastIndex: index,
      path: path,
      text: char,
      pathBoundingBox: pathBoundingBox,
      chilePath,
    };
  }

  return {
    pathParts: context.pathParts,
    pathPartsAlignTransform: context.pathPartsAlignTransform,
    pathPartsTransform: context.pathPartsTransform,
    position,
    svgSize,
    lineHeight,
    isVertical,
    domBoxSize: {
      width: config.MaxWidth,
      height: config.MaxHeight,
    },
    hasCnChar: true,
  };
}
function isChinese(char) {
  if (char === " ") return true;
  // [\u4e00-\u9fff]：匹配中文汉字。
  // [\u3000-\u303f]：匹配CJK符号和标点（例如全角逗号、句号等）。
  // [\uff00-\uffef]：匹配全角字符和标点符号（例如全角括号、感叹号、引号等）。
  return /[\u4e00-\u9fff\u3000-\u303f\uff00-\uffef]/.test(char);
}

function isEnglish(char) {
  // 英文字母的 Unicode 范围是 A-Z 或 a-z
  return /^[A-Za-z]$/.test(char);
}
function hasCnWords(str) {
  return /[\u4e00-\u9fa5]/.test(str);
}
//   截止判断
function isEnd(char) {
  // 1. 遇到中文
  if (isChinese(char)) return true;
  // 2. 字符结束
  if (char === undefined || char === null) return true;
}
