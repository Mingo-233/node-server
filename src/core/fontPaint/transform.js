function transform() {
  // TODO: scale
  // TODO: pageTranslate
}
