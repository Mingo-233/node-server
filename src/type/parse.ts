import type { Path } from "opentype.js";
enum IFontShape {
  glyph = 1,
  letter = 2,
}

export interface IPosition {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
}
interface IPathPart {
  path: Path;
  translate: string;
  alignTranslate: string;
}
export type IPathCollection = IPathPart[];

export interface IGenSvgConfig {
  position: IPosition;
  isVertical: boolean;
  domBoxSize: {
    width: number;
    height: number;
  };
  DPI: number;
  hasSymbolChar: boolean;
  pageTranslate: string;
}
