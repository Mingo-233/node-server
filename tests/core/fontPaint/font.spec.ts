import { beforeAll, describe, expect, it } from "vitest";
import { getTextPaths } from "@/core/fontPaint/parse";
import { genSvgCode } from "@/core/fontPaint/generate";
import wawoff from "wawoff2";
import opentype from "opentype.js";
import path from "path";
import fs from "fs";
// vitest 在解析opentype.js的时候，会对内容重写。 而opentype.js内部做了 object.frozen，导致会报错。
describe("字体转曲测试", () => {
  let fontApp: any = null;
  beforeAll(async () => {
    try {
      const fontPath = path.join(
        __dirname,
        "../../../example/fonts/font.woff2"
      );
      const buffer = await fs.promises.readFile(fontPath);
      const data = await wawoff.decompress(buffer);
      const arrayBuffer = data.buffer.slice(
        data.byteOffset,
        data.byteOffset + data.byteLength
      );
      fontApp = opentype.parse(arrayBuffer);
    } catch (error) {
      console.error("Error reading file:", error);
      throw error; // 重新抛出错误以便 Vitest 能够报告它
    }
  });
  it("英文 /n换行情况", async () => {
    const {
      pathParts: paths,
      pathPartsTransform,
      pathPartsAlignTransform,
      position,
      svgSize,
      lineHeight,
      isVertical,
      domBoxSize,
      hasCnChar,
    } = getTextPaths(fontApp, {
      text: "hello\nworld",
      fontSize: 30,
      textAlign: "left",
      vertical: 0,
      MaxWidth: 200,
      MaxHeight: 80,
    });
    expect(Object.keys(paths).length).toBe(2);
    const svgDom = genSvgCode(paths, {
      position,
      svgSize,
      lineHeight,
      isVertical,
      pathPartsTransform,
      pathPartsAlignTransform,
      domBoxSize,
      hasCnChar,
    });
    // 生成快照
    expect(svgDom).toMatchSnapshot();
  });
});
