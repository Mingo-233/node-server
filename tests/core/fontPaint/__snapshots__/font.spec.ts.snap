// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`字体转曲测试 > 英文 /n换行情况 1`] = `
"<svg xmlns="http://www.w3.org/2000/svg" id="preview" width="200mm" height="80mm" viewBox="1.98 -21 200 80" transform="" fill="red">
      <g >
      
                  <g transform="">
                      <path transform="translate(0,NaN)"  d="M9.450000-15.390000C7.350000-15.390000 5.700000-14.610000 4.590000-12.840000L4.590000-21L1.980000-21L1.980000 0L4.590000 0L4.590000-7.950000C4.590000-11.490000 6.540000-12.900000 8.940000-12.900000C11.250000-12.900000 12.660000-11.520000 12.660000-9.060000L12.660000 0L15.270000 0L15.270000-9.210000C15.270000-13.050000 12.960000-15.390000 9.450000-15.390000ZM20.730000-6.300000L33.300000-6.300000C33.360000-6.690000 33.390000-7.080000 33.390000-7.470000C33.390000-11.670000 30.450000-15.390000 25.920000-15.390000C21.240000-15.390000 18.060000-11.940000 18.060000-7.500000C18.060000-3 21.270000 0.390000 26.100000 0.390000C29.100000 0.390000 31.380000-0.960000 32.700000-3L30.480000-4.260000C29.730000-3.030000 28.230000-2.070000 26.160000-2.070000C23.370000-2.070000 21.240000-3.600000 20.730000-6.300000ZM25.920000-12.930000C28.170000-12.930000 30.300000-11.550000 30.750000-8.580000L20.730000-8.580000C21.150000-11.190000 23.100000-12.930000 25.920000-12.930000ZM36.390000 0L39 0L39-21.900000L36.390000-21.900000ZM42.990000 0L45.600000 0L45.600000-21.900000L42.990000-21.900000ZM56.640000 0.390000C61.020000 0.390000 64.560000-3.060000 64.560000-7.500000C64.560000-11.940000 61.020000-15.390000 56.640000-15.390000C52.260000-15.390000 48.750000-11.940000 48.750000-7.500000C48.750000-3.060000 52.260000 0.390000 56.640000 0.390000ZM56.640000-2.160000C53.670000-2.160000 51.360000-4.470000 51.360000-7.500000C51.360000-10.530000 53.670000-12.840000 56.640000-12.840000C59.640000-12.840000 61.950000-10.530000 61.950000-7.500000C61.950000-4.470000 59.640000-2.160000 56.640000-2.160000Z"/>
                  </g>
                  
                  <g transform="">
                      <path transform="translate(0,NaN)"  d="M19.650000-15L16.350000-3.570000L12.810000-15L10.290000-15L6.750000-3.600000L3.450000-15L0.720000-15L5.370000 0L8.040000 0L11.550000-11.100000L15.060000 0L17.730000 0L22.380000-15ZM31.530000 0.390000C35.910000 0.390000 39.450000-3.060000 39.450000-7.500000C39.450000-11.940000 35.910000-15.390000 31.530000-15.390000C27.150000-15.390000 23.640000-11.940000 23.640000-7.500000C23.640000-3.060000 27.150000 0.390000 31.530000 0.390000ZM31.530000-2.160000C28.560000-2.160000 26.250000-4.470000 26.250000-7.500000C26.250000-10.530000 28.560000-12.840000 31.530000-12.840000C34.530000-12.840000 36.840000-10.530000 36.840000-7.500000C36.840000-4.470000 34.530000-2.160000 31.530000-2.160000ZM45.150000-12.480000L45.150000-15L42.540000-15L42.540000 0L45.150000 0L45.150000-7.950000C45.150000-11.490000 47.580000-12.630000 49.830000-12.540000L49.830000-15.270000C47.940000-15.270000 46.110000-14.580000 45.150000-12.480000ZM52.380000 0L54.990000 0L54.990000-21.900000L52.380000-21.900000ZM71.400000-21L71.400000-12.420000C70.200000-14.250000 68.280000-15.390000 65.730000-15.390000C61.560000-15.390000 58.140000-11.940000 58.140000-7.500000C58.140000-3.060000 61.560000 0.390000 65.730000 0.390000C68.280000 0.390000 70.200000-0.750000 71.400000-2.580000L71.400000 0L74.010000 0L74.010000-21ZM66.060000-2.130000C63.060000-2.130000 60.750000-4.440000 60.750000-7.500000C60.750000-10.560000 63.060000-12.870000 66.060000-12.870000C69.090000-12.870000 71.400000-10.560000 71.400000-7.500000C71.400000-4.440000 69.090000-2.130000 66.060000-2.130000Z"/>
                  </g>
                  
      </g>
      </svg>"
`;
